### ISSUE TITLE
Optionally not start the daemon

### ISSUE THREAD

**Issue by romw** [BOINC/boinc#79]
**Reported by KSMarksPsych on 17 Jun 37299993 08:00 UTC**
Posted by: 
Date: 7:19 AM 01-06-2006
Add a command line flag (--no-daemon) or a persistant menu option in the Manager to allow people to optionally have the manager not fire up the daemon if it isn't there. Many users (myself included) don't want to run the daemon on things such as our portable computers, but do want to admin other machines from our portables

Migrated-From: http://boinc.berkeley.edu/trac/ticket/79

**Comment by AenBleidd** [comment]
Fixed.

- New commanline option -nd (short) and --no-daemon (long).
- New option in Preferences Dialog "Run daemon?"
So after Manager start daemon will be run only if commandline option --no-daemon is off and option "Run daemon?" is on. Otherwise daemon will not be run.

**Comment by AenBleidd** [comment]
This ticket was done 2 years ago and was released 2 or more versions ago.
@Ageless93, could you please test it and close this issue?
Thanks

**Comment by Ageless93** [comment]
I have tested the GUI option "Run the client?" from BOINC 7.8.3 with Manager 7.9.0
Doing Options->Other options...->uncheck "Run the client?"->Save.
Exit BOINC Manager, tell it to exit running tasks.
Restart BOINC Manager. This only starts the manager, as it should.

To fix at this point.
Options->Other options->check "Run the client?"->Save.
Exit BOINC Manager.
Restart BOINC Manager. This starts both the client and the manager. 

For completeness, but because this is only in master, not in the 7.8 branch:
Opened command line window.
Doing boinc.exe -nd gives me Unknown option: -nd
Doing boinc.exe --nd gives me Unknown option: --nd
Doing boinc.exe --no-daemon gives me Unknown option: --no-daemon

Then did boincmgr.exe -nd
This starts BOINC Manager, disconnected from a client. However, Options->Other options shows "Run the client?" as checked. 
Doing boincmgr.exe -no_daemon starts BOINC Manager, disconnected from a client. Trying to get to Other options or even to exit BOINC Manager, slowed everything right down and leaves the manager in an unresponsive way on the desktop. Essentially that crashed it. Had to exit it via Windows Task Manager.  

Both options should be with two hyphens, --nd and --no-daemon.

I'll build a client out of master tomorrow and test it there then.

**Comment by AenBleidd** [comment]
@Ageless93,

> This starts BOINC Manager, disconnected from a client. However, Options->Other options shows "Run the client?" as checked.

This is by design: commandline option is one time use, while Option in Settings is for every run (command line option has bigger priority).

> Doing boincmgr.exe -no-daemon starts BOINC Manager, disconnected from a client. Trying to get to Other options or even to exit BOINC Manager, slowed everything right down and leaves the manager in an unresponsive way on the desktop. Essentially that crashed it. Had to exit it via Windows Task Manager.

I'll check it. When I implemented this two years ago all was ok. Maybe something have changed.

> Both options should be with two hyphens, --nd and --no-daemon.

Are you sure? Do other 'short' options have two hyphens too?

**Comment by Ageless93** [comment]
@AenBleidd, yes all command line attributes start with double hyphens:

C:\Users\Ageless>boinc.exe --help
The command-line options for boinc.exe are intended for debugging.
The recommended command-line interface is a separate program,'boinccmd'.
Run boinccmd in the same directory as boinc.exe.

Usage: boinc.exe [options]
```
    --abort_jobs_on_exit           when client exits, abort and report jobs
    --allow_remote_gui_rpc         allow remote GUI RPC connections
    --allow_multiple_clients       allow >1 instances per host
    --attach_project <URL> <key>   attach to a project
    --check_all_logins             for idle detection, check remote logins too
    --daemon                       run as daemon (Unix)
    --detach_console               detach from console (Windows)
    --detach_project <URL>         detach from a project
    --dir <path>                   use given dir as BOINC home
    --exit_after_app_start N       exit N seconds after an app starts
    --exit_after_finish            exit right after finishing a job
    --exit_before_start            exit right before starting a job
    --exit_before_upload           exit right before starting an upload
    --exit_when_idle               exit when there are no results
    --fetch_minimal_work           fetch only 1 job per device
    --file_xfer_giveup_period N    give up on file xfers after N sec
    --gui_rpc_port <port>          port for GUI RPCs
    --gui_rpc_unix_domain          use Unix domain for GUI RPCs
    --help                         show options
    --insecure                     disable app sandboxing (Unix)
    --master_fetch_interval N      limiting period of master retry
    --master_fetch_period N        reload master URL after N RPC failures
    --master_fetch_retry_cap N     exponential backoff limit
    --no_gpus                      don't check for GPUs
    --no_gui_rpc                   don't allow GUI RPC, don't make socket
    --no_info_fetch                don't fetch project list or client version info
    --no_priority_change           run apps at same priority as client
    --pers_giveup N                giveup time for persistent file xfer
    --pers_retry_delay_max N       max for file xfer exponential backoff
    --pers_retry_delay_min N       min for file xfer exponential backoff
    --redirectio                   redirect stdout and stderr to log files
    --reset_project <URL>          reset (clear) a project
    --retry_cap N                  exponential backoff limit
    --run_cpu_benchmarks           run the CPU benchmarks
    --run_by_updater               set by updater
    --sched_retry_delay_max N      max for RPC exponential backoff
    --sched_retry_delay_min N      min for RPC exponential backoff
    --show_projects                show attached projects
    --skip_cpu_benchmarks          don't run CPU benchmarks
    --start_delay X                delay starting apps for X secs
    --suppress_net_info            don't send network addrs to server
    --unsigned_apps_ok             allow unsigned apps (for testing)
    --update_prefs <URL>           contact a project to update preferences
    --version                      show version info
```
attributes going through BOINC Manager either start with / for short, or -- for long:
```
C:\Users\Ageless>boincmgr.exe /help
Usage: boincmgr [/h] [--verbose] [/a] [/s] [/n <str>] [/g <str>] [/p <str>] [/b
<str>] [/i] [/c] [/m] [/nd]
  /h, --help                    show this help message
  --verbose                     generate verbose log messages
  /a, --autostart               BOINC Manager was started by the operating syste
m automatically
  /s, --systray                 Startup BOINC so only the system tray icon is vi
sible
  /n, --namehost=<str>          Host name or IP address
  /g, --gui_rpc_port=<str>      GUI RPC port number
  /p, --password=<str>          Password
  /b, --boincargs=<str>         Startup BOINC with these optional arguments
  /i, --insecure                disable BOINC security users and permissions
  /c, --checkskins              set skin debugging mode to enable skin manager e
rror messages
  /m, --multiple                multiple instances of BOINC Manager allowed
  /nd, --no-daemon              Don't run the client
```

**Comment by Ageless93** [comment]
Built boinc.exe out of Master
C:\Users\Ageless>boinc.exe --version
7.9.0 windows_x86_64

Doing boinc.exe -nd gives me Unknown option: -nd
Doing boinc.exe --nd gives me Unknown option: --nd
Doing boinc.exe -no-daemon gives me Unknown option: -no-daemon
Doing boinc.exe --no-daemon gives me Unknown option: --no-daemon

So, while it works from the manager, it doesn't seem to work from the client. Keeping this issue open for now.

**Comment by AenBleidd** [comment]
@Ageless93,
Sorry maybe this is a misunderstanding here but the description of the ticket says ha have an option to run Manager without firing up a daemon (client). So we we need to have such option for client (boinc.exe)? What this option should do in this case? Run boinc.exe and terminate it immediately?
I added an option to Manager only and this is by design.

**Comment by Ageless93** [comment]
Ah, you're right. I'm not awake yet. :-)

That leaves it with the manager. 
boincmgr.exe /nd or no-daemon, then Options->Other options should show "Run the client?" as unchecked. Because that's how you started the manager. If you want to start the client at a next manager start-up, you'll need to check "Run the client?", just as you would have to if you did all thise via the GUI. 

Furthermore, and I think this was referenced in another issue as well, the other attributes for client and manager, with more than one word (e.g. gui_rpc_port), use underscores between words, not hyphens. I think we should use that everywhere. So no_daemon instead of no-daemon

**Comment by CharlieFenton** [comment]
Actually, both single and double hyphens are accepted, at least on some platforms. In client/cs_cmdline.cpp line 116 is:
#define ARG(S) ARGX2("-"#S, "--"#S)
so both are accepted for the client, but apparently / is not.

For the Manager, the optional use of the slash / is specific to MS Windows. The wxWidgets documentation for wxCmdLineParser says:
> First global option is the support for long (also known as GNU-style) options. The long options are the ones which start with two dashes and look like "--verbose", i.e. they generally are complete words and not some abbreviations of them. As long options are used by more and more applications, they are enabled by default, but may be disabled with DisableLongOptions().
> 
> Another global option is the set of characters which may be used to start an option (otherwise, the word on the command line is assumed to be a parameter). Under Unix, "-" is always used, but Windows has at least two common choices for this: "-" and "/". Some programs also use "+". The default is to use what suits most the current platform, but may be changed with SetSwitchChars() method.


On the Mac (since mac OS X is built on top of UNIX), both "BOINCManager -help" and  "BOINCManager --help" give this:
Usage: BOINCManager [-h] [--verbose] [-a] [-s] [-n <str>] [-g <str>] [-p <str>] [-b <str>] [-i] [-c] [-m] [-nd]
  -h, --help              	show this help message
  --verbose               	generate verbose log messages
  -a, --autostart         	BOINC Manager was started by the operating system automatically
  -s, --systray           	Startup BOINC so only the system tray icon is visible
  -n, --namehost=<str>    	Host name or IP address
  -g, --gui_rpc_port=<str>	GUI RPC port number
  -p, --password=<str>    	Password
  -b, --boincargs=<str>   	Startup BOINC with these optional arguments
  -i, --insecure          	disable BOINC security users and permissions
  -c, --checkskins        	set skin debugging mode to enable skin manager error messages
  -m, --multiple          	multiple instances of BOINC Manager allowed
  -nd, --no-daemon        	Don't run the client

**Comment by AenBleidd** [comment]
@Ageless93,

> That leaves it with the manager.
boincmgr.exe /nd or no-daemon, then Options->Other options should show "Run the client?" as unchecked. Because that's how you started the manager. If you want to start the client at a next manager start-up, you'll need to check "Run the client?", just as you would have to if you did all thise via the GUI.

Sorry but I do not agree with you. From my POV command line argument must not change user defined settings. E.g. you want always to start client on Manager start but occasionally you need just to run Manager without firing up client so you start Manager from command line with '--no-daemon' parameter. But the actual parameter should not be changed because user will not be even notified that the parameter was changed and he or she need to change it if it planned to start client with Manager in future.

**Comment by CharlieFenton** [comment]
I agree with @AenBleidd.

**Comment by AenBleidd** [comment]
@TheAspens, could you please take a look at this ticket?
It seems to be fixed more than 2 years ago.
The corresponded code was added to master branch and already changed by @davidpanderson 
From my POV this issue can be closed as resolved and possibly released in 7.10 release.

**Comment by AenBleidd** [comment]
Fixed via #1374
